File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/WebActions/ActionInvoker.cs
﻿using System;
using JDI_Commons;
using Epam.JDI.Core.Logging;
using Epam.JDI.Core.Settings;
using JDI_Web.Selenium.Base;
using JDI_Web.Selenium.Elements.Base;
using static Epam.JDI.Core.ExceptionUtils;

namespace JDI_Web.Selenium.Elements.WebActions
{
    public class ActionInvoker
    {
        private readonly WebBaseElement _element;
        public static ActionScenrios ActionScenrios = new ActionScenrios();

        public ActionInvoker(WebBaseElement element)
        {
            JDISettings.NewTest();
            _element = element;
        }

        public TResult DoJActionResult<TResult>(string actionName, Func<WebBaseElement, TResult> action,
            Func<TResult, string> logResult = null, LogLevels level = LogLevels.Info)
        {
            return ActionWithException(() =>
            {
                ProcessDemoMode();
                return ActionScenrios.SetElement(_element).ResultScenario(actionName, action, logResult, level);
            }, ex => $"Failed to do '{actionName}' action. Reason: {ex}");
        }
        
        public void DoJAction(string actionName, Action<WebBaseElement> action, LogLevels level = LogLevels.Info)
        {
            TimerExtensions.ForceDone(() => {
                ProcessDemoMode();
                ActionScenrios.SetElement(_element).ActionScenario(actionName, action, level);
            });
        }

        public void ProcessDemoMode()
        {
            if (!JDISettings.IsDemoMode) return;
            if (_element is WebElement)
                ((WebElement)_element).Highlight(JDISettings.HighlightSettings);
        }
    }
}

File: jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winnium/actions/ActionInvoker.java
package com.epam.jdi.uitests.win.winnium.actions;

import com.epam.commons.linqinterfaces.JAction;
import com.epam.jdi.uitests.core.logger.LogLevels;

import java.util.function.Function;
import java.util.function.Supplier;

import static com.epam.jdi.uitests.core.logger.LogLevels.INFO;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

public class ActionInvoker {
    private ActionScenarios actionScenrios = new ActionScenarios();

    public final <TResult> TResult doJActionResult(String actionName, Supplier<TResult> action, String elementToString) {
        return doJActionResult(actionName, action, null, INFO, elementToString);
    }

    public final <TResult> TResult doJActionResult(String actionName, Supplier<TResult> action, LogLevels level,
                                                   String elementToString) {
        return doJActionResult(actionName, action, null, level, elementToString);
    }

    public final <TResult> TResult doJActionResult(String actionName, Supplier<TResult> action,
                                                   Function<TResult, String> logResult, LogLevels level,
                                                   String elementToString) {
        try {
            return actionScenrios.resultScenario(actionName, action, logResult, level, elementToString);
        } catch (Exception | Error ex) {
            throw exception("Failed to do '%s' action. Reason: %s", actionName, ex);
        }
    }

    public final void doJAction(String actionName, JAction action, String elementToString) {
        doJAction(actionName, action, INFO, elementToString);
    }

    public final void doJAction(String actionName, JAction action, LogLevels level, String elementToString) {
        actionScenrios.actionScenario(actionName, action, level, elementToString);
    }

}
